### PR TITLE
Heightmanager unit test

### PR DESCRIPTION
**Pull request by eric998101**
add heightmanager unit test

### CODE REVIEW

File: committee/heightmanager_test.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+package committee
+
+import (
+	"strconv"
+	"testing"
+	"time"
+)
+
+var baseTime = time.Unix(int64(1546272000), 0) // 2019-01-01 00:00:00
+var hm *heightManager
```
Comment (raullenchai) [inline committee/heightmanager_test.go]: Let's have an hm in each test, so there is no dependency between tests.

File: committee/heightmanager_test.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+package committee
+
+import (
+	"strconv"
+	"testing"
+	"time"
+)
+
+var baseTime = time.Unix(int64(1546272000), 0) // 2019-01-01 00:00:00
```
Comment (raullenchai) [inline committee/heightmanager_test.go]: const

File: committee/heightmanager_test.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+var hm *heightManager
+
+func TestNewHeightManager(t *testing.T) {
+	hm = newHeightManager()
+	if hm == nil {
```
Comment (CoderZhi) [inline committee/heightmanager_test.go]: require.NotNil(t, hm)

File: committee/heightmanager_test.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	for i := 0; i < 5; i++ {
+		ts := baseTime.Add(time.Second * time.Duration(15*i)) // increments in 15 seconds
+		name := strconv.Itoa(i) + "-" + strconv.FormatInt(ts.Unix(), 10)
+		te := test{name, hm, args{uint64(i), ts}, false}
+		tests = append(tests, te)
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := tt.m.add(tt.args.height, tt.args.ts); (err != nil) != tt.wantErr {
```
Comment (CoderZhi) [inline committee/heightmanager_test.go]: err := tt.m.add(tt.args.height, tt.args.ts)
if tt.wantErr {
  require.Error(err)
} else {
  require.NoError(err)
}

In general, we don't need to make it so complicated by defining struct.

File: committee/heightmanager_test.go (hunk 3 above)
Comment (eric998101) [reply committee/heightmanager_test.go]: got github.com/stretchr/testify/require, nice library, figure it out now.

File: committee/heightmanager_test.go (hunk 4 above)
Comment (eric998101) [reply committee/heightmanager_test.go]: it's auto-generated code, actually, not very confused.

File: committee/heightmanager_test.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+	tests := []struct {
+		name    string
+		m       *heightManager
+		args    args
+		wantErr bool
+	}{
+		{"0-1546272000", hm, args{0, baseTime}, true},
+		{"4-1546272060", hm, args{4, baseTime.Add(time.Second * 60)}, true},
+		{"5-1546272060", hm, args{5, baseTime.Add(time.Second * 60)}, true},
```
[the remark this answers is not in the document]
Comment (eric998101) [reply committee/heightmanager_test.go]: update to the following 9 cases to cover all situation
// h < latest, t < latest
// h < latest, t = latest
// h < latest, t > latest
// h = latest, t < latest
// h = latest, t = latest
// h = latest, t > latest
// h > latest, t < latest
// h > latest, t = latest
// h > latest, t > latest  only this case is valid

File: committee/heightmanager_test.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+	tests := []struct {
+		name string
+		m    *heightManager
+		args args
+		want uint64
+	}{
+		{"1546271999", hm, args{baseTime.Add(time.Second * -1)}, 0},
+		{"1546272001", hm, args{baseTime.Add(time.Second * 1)}, 0},
+		{"1546272035", hm, args{baseTime.Add(time.Second * 35)}, 2},
```
[the remark this answers is not in the document]
Comment (eric998101) [reply committee/heightmanager_test.go]: update to test -30, -1, 0, 1, 14, 15, 16, 29, 30, 31, 59, 60, 61, 90

File: committee/heightmanager_test.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+		})
+	}
+}
+
+func TestLastestHeight(t *testing.T) {
+	baseTime := time.Unix(int64(1546272000), 0) // 2019-01-01 00:00:00
+	hm := newHeightManager()
+	for i := 0; i < 5; i++ {
+		ts := baseTime.Add(time.Second * time.Duration(15*i)) // increments in 15 seconds
+		hm.add(uint64(i), ts)
```
[the remark this answers is not in the document]
Comment (eric998101) [reply committee/heightmanager_test.go]: That's more reasonable, done.

**Pull request closed without merging**